draft/ZEP0002.md

```diff
@@ -0,0 +1,285 @@
+
+Image datasets in the biomedical, geo-science and general data-science communities are steadily growing. In recent years, peta-scale datasets have been made publicly available. Although peta-scale datasets are still infrequent, tera-scale datasets are acquired routinely and rapidly grow in size further.
+
+Currently, these large-scale datasets are made available through tool-specific formats, such as neuroglancer-precomputed or webknossos-wrap, which support sharding. While the data is accessible and explorable in individual software tools, wider interoperability with other software tools is limited. Whereas it has been possible to convert smaller datasets to work in different tools, it becomes infeasible for large-scale datasets. The choice of tool-specific formats is in part a result of the lack of a standard format that makes storage and processing of very large datasets practical. Providing sharding in Zarr would allow Zarr to become the standard format also for tera- and peta-scale array data and future-proofing it as a standard for the requirements of growing data sizes in the scientific community.
+
+
+## Backward Compatibility
+
+This ZEP introduces a Zarr extension for Zarr version 3. The sharding
+extension is optional for data producers, and only necessary when reading
+arrays that are written using the sharding extension. It corresponds
+to a minor version change in semantic versioning, adding functionality
+in a backwards compatible manner.
+
+
+## Detailed description
+
+
+### Combining regular blocks of chunks in a shard
+
+To add sharding to Zarr, data that was stored previously as separate objects must be combined.
+Different strategies can be considered when sharding array data, such as:
+1. sharding across single dimensions,
+2. sharding by regular n-dimensional blocks of data, and
+3. sharding random combinations of chunks via hashing.
+
+All of those strategies only shard the array-data, both as metadata must first be read without
+sharding, so that sharding itself can be configured in the metadata, and also since the relevant
+parts of the data is the array data itself.
+
+Strategy 2 is a strict super-set of strategy 1, as sharding across a single dimension can
+be produced by sharding with n-dimensional blocks where all blocks consist of single chunks
+except for the selected dimension.
+
+Strategy 3 allows to balance loads better between shards than the other two strategies. However,
+sharding is not meant to distribute loads in this case, since all shards will mainly be served
+via the same underlying storage, but it instead reduces the number of stored files or objects.
+
+Comparing strategy 3 and 2, in opposite to the load balancing, it is more important to co-locate
+data that us usually accessed together or in short succession. This is optimized by combining
+multiple chunks that are close to each other, what is supported best by strategy 2, using
+regular n-dimensional blocks of chunks as sharding units.
```

Comment:
this paragraph could be a bit clearer. maybe:

```suggestion
When comparing strategy 3 and 2, it's important to note that it is less important to do load balancing than to co-locate
data that is usually accessed together or in short succession. The latter can be achieved by combining multiple chunks that are close to each other, which is supported best by strategy 2, using
regular n-dimensional blocks of chunks as sharding units.
```

Reply:
I don't know if we have to be absolute about 2 being better than 3. For this proposal, we implement 2. Another storage transformer extension could choose to implement 3, if the use cases call for it.

Reply:
True, fair enough. I restructured the arguments a bit and think this is now more balanced.